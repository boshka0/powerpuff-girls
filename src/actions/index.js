import { SET_SHOW, SET_EPISODES, SET_EPISODE } from '../constants';

export const setAllEpisodes = episodes => ({
  type: SET_EPISODES,
  payload: episodes,
});

export const setShow = show => ({
  type: SET_SHOW,
  payload: show,
});

export const setEpisode = episode => ({
  type: SET_EPISODE,
  payload: episode,
});
