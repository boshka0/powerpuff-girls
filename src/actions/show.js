import { SHOW } from '../constants';

export const loadShow = () => ({
  type: SHOW.LOAD,
});

export const setShow = episode => ({
  type: SHOW.LOAD_SUCCESS,
  payload: episode,
});

export const setShowError = error => ({
  type: SHOW.LOAD_FAIL,
  payload: error,
});
