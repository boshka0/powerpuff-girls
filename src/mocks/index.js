export const episodes = [{
    id: 657308,
    url: 'http://www.tvmaze.com/episodes/657308/the-powerpuff-girls-1x01-escape-from-monster-island',
    name: 'Escape from Monster Island',
    season: 1,
    number: 1,
    airdate: '2016-04-04',
    airtime: '18:00',
    airstamp: '2016-04-04T22:00:00+00:00',
    runtime: 11,
    image: {
      medium: 'http://static.tvmaze.com/uploads/images/medium_landscape/53/132617.jpg',
      original: 'http://static.tvmaze.com/uploads/images/original_untouched/53/132617.jpg'
    },
    summary: "<p>Bubbles wins two tickets to a concert and has to pick between Blossom and Buttercup to go with her. Meanwhile, the Mayor's plane goes down over Monster Island and it's up to the girls to rescue him.</p>",
    _links: {
      self: {
        href: 'http://api.tvmaze.com/episodes/657308'
      }
    }
  },
  {
    id: 657309,
    url: 'http://www.tvmaze.com/episodes/657309/the-powerpuff-girls-1x02-princess-buttercup',
    name: 'Princess Buttercup',
    season: 1,
    number: 2,
    airdate: '2016-04-04',
    airtime: '18:00',
    airstamp: '2016-04-04T22:00:00+00:00',
    runtime: 15,
    image: {
      medium: 'http://static.tvmaze.com/uploads/images/medium_landscape/53/132618.jpg',
      original: 'http://static.tvmaze.com/uploads/images/original_untouched/53/132618.jpg'
    },
    summary: '<p>Buttercup becomes enamored with a team of roller derby girls called the Derbytantes. When she starts spending all of her time with them, Princess Morbucks seeks to be her replacement in the Powerpuff Girls.</p>',
    _links: {
      self: {
        href: 'http://api.tvmaze.com/episodes/657309'
      }
    }
  }
];

export const show = {
  id: 657309,
  url: 'http://www.tvmaze.com/episodes/657309/the-powerpuff-girls-1x02-princess-buttercup',
  name: 'Princess Buttercup',
  season: 1,
  number: 2,
  airdate: '2016-04-04',
  airtime: '18:00',
  airstamp: '2016-04-04T22:00:00+00:00',
  runtime: 15,
  image: {
    medium: 'http://static.tvmaze.com/uploads/images/medium_landscape/53/132618.jpg',
    original: 'http://static.tvmaze.com/uploads/images/original_untouched/53/132618.jpg'
  },
  summary: '<p>Buttercup becomes enamored with a team of roller derby girls called the Derbytantes. When she starts spending all of her time with them, Princess Morbucks seeks to be her replacement in the Powerpuff Girls.</p>',
  _links: {
    self: {
      href: 'http://api.tvmaze.com/episodes/657309'
    }
  }
};
