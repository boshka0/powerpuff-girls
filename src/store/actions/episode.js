import { EPISODE } from '../constants';

export const loadEpisode = id => ({
  type: EPISODE.LOAD,
  payload: id,
});

export const setEpisode = episode => ({
  type: EPISODE.LOAD_SUCCESS,
  payload: episode,
});

export const setEpisodeError = error => ({
  type: EPISODE.LOAD_FAIL,
  payload: error,
});
