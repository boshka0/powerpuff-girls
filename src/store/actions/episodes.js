import { EPISODES } from '../constants';

export const loadEpisodes = () => ({
  type: EPISODES.LOAD,
});

export const setEpisodes = episode => ({
  type: EPISODES.LOAD_SUCCESS,
  payload: episode,
});

export const setEpisodesError = error => ({
  type: EPISODES.LOAD_FAIL,
  payload: error,
});
