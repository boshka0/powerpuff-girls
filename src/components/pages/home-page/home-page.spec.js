import React from 'react';
import configureMockStore from 'redux-mock-store';
import { HashRouter as Router } from 'react-router-dom';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import { Provider } from 'react-redux';

import { loadShow } from '../../../store/actions/show';
import { loadEpisodes } from '../../../store/actions/episodes';
import { episodes, show } from '../../../mocks';

import HomePage from './index';

describe('episode-page', () => {
  it('should render correctly home-page if episodes and show info are available', () => {
    const episodesMockStore = configureMockStore()({
      episodes: {
        data: episodes,
        loading: false,
      },
      show: {
        data: show,
        loading: false
      }
    });

    const wrapper = mount(
      <Provider store={episodesMockStore}>
        <Router>
          <HomePage />
        </Router>
      </Provider>
    );
    expect(renderToJson(wrapper.render())).toMatchSnapshot();
  });

  it('should render errors if there was an error while fetching data', () => {
    const errorMockStore = configureMockStore()({
      episodes: {
        error: 'API call failed',
        loading: false,
      },
      show: {
        error: 'API call failed',
        loading: false,
      }
    });

    const wrapper = mount(
      <Provider store={errorMockStore}>
        <Router>
          <HomePage />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.error')).toHaveLength(2);
    expect(renderToJson(wrapper.render())).toMatchSnapshot();
  });

  it('should redirect user to episode page on clicking episode preview', () => {
    const episodesMockStore = configureMockStore()({
      episodes: {
        data: episodes,
        loading: false,
      },
      show: {
        data: show,
        loading: false
      }
    });

    const wrapper = mount(
      <Provider store={episodesMockStore}>
        <Router>
          <HomePage />
        </Router>
      </Provider>
    );
    wrapper.find('.about-episodes__link').first().simulate('click', { button: 0 });
    expect(global.location.hash).toEqual(`#/episodes/${episodes[0].id}`);
  });

  it('should dispatch EPISODES_LOAD and SHOW_LOAD actions', () => {
    const emptyMockStore = configureMockStore()({});
    const wrapper = mount(
      <Provider store={emptyMockStore}>
        <Router>
          <HomePage />
        </Router>
      </Provider>
    );

   expect(wrapper.props().store.getActions()).toContainEqual(loadEpisodes());
   expect(wrapper.props().store.getActions()).toContainEqual(loadShow());
  });
});
