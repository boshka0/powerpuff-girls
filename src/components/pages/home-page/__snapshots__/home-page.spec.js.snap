// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`episode-page should render correctly home-page if episodes and show info are available 1`] = `
<div
  class="about"
>
  <div
    class="about-show"
  >
    <h1
      class="about-show__title"
    >
      Princess Buttercup
    </h1>
    <div
      class="about-show__info"
    >
      <img
        alt="Princess Buttercup"
        class="about-show__image"
        src="http://static.tvmaze.com/uploads/images/medium_landscape/53/132618.jpg"
      />
      <div
        class="about-show__description"
      >
        <p>
          Buttercup becomes enamored with a team of roller derby girls called the Derbytantes. When she starts spending all of her time with them, Princess Morbucks seeks to be her replacement in the Powerpuff Girls.
        </p>
      </div>
    </div>
  </div>
  <div
    class="about-episodes"
  >
    <h2
      class="about-episodes__title"
    >
      Episode list:
    </h2>
    <div
      class="about-episodes__container"
    >
      <a
        class="about-episodes__link"
        href="#/episodes/657308"
      >
        <div
          class="episode-preview"
        >
          <img
            alt="Escape from Monster Island"
            class="episode-preview__image"
            src="http://static.tvmaze.com/uploads/images/medium_landscape/53/132617.jpg"
          />
          <div
            class="episode-preview__title"
          >
            <p
              class="episode-preview__title-count"
            >
              Season 1. Episode 1
            </p>
            <p>
              Escape from Monster Island
            </p>
          </div>
        </div>
      </a>
      <a
        class="about-episodes__link"
        href="#/episodes/657309"
      >
        <div
          class="episode-preview"
        >
          <img
            alt="Princess Buttercup"
            class="episode-preview__image"
            src="http://static.tvmaze.com/uploads/images/medium_landscape/53/132618.jpg"
          />
          <div
            class="episode-preview__title"
          >
            <p
              class="episode-preview__title-count"
            >
              Season 1. Episode 2
            </p>
            <p>
              Princess Buttercup
            </p>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`episode-page should render errors if there was an error while fetching data 1`] = `
<div
  class="about"
>
  <div
    class="error"
  >
    Sorry, some error occurred while preparing the response for show info
  </div>
  <div
    class="error"
  >
    Sorry, some error occurred while preparing the response for episodes list
  </div>
</div>
`;
